File: server-go/viewport_test.go
package main

import "testing"

func TestPanUpDown(t *testing.T) {
	if got := panViewport(View{Row: 5, Col: 0, Follow: true}, "up"); got.Row != 4 || got.Follow {
		t.Fatalf("up %+v", got)
	}
	if got := panViewport(View{Row: 5, Col: 0, Follow: false}, "down"); got.Row != 6 || got.Follow {
		t.Fatalf("down %+v", got)
	}
}

func TestPanLeftRight(t *testing.T) {
	if panViewport(View{Row: 0, Col: 10, Follow: false}, "left").Col != 2 {
		t.Fatal("left")
	}
	if panViewport(View{Row: 0, Col: 0, Follow: false}, "left").Col != 0 {
		t.Fatal("left floor")
	}
	if panViewport(View{Row: 0, Col: 0, Follow: false}, "right").Col != colStep {
		t.Fatal("right")
	}
}

func TestPanFloorAndUnknown(t *testing.T) {
	if panViewport(View{Row: 0, Col: 0, Follow: false}, "up").Row != 0 {
		t.Fatal("row floor")
	}
	got := panViewport(View{Row: 2, Col: 2, Follow: true}, "zzz")
	if got.Row != 2 || got.Col != 2 || !got.Follow {
		t.Fatalf("unknown %+v", got)
	}
}

func TestWindowLines(t *testing.T) {
	grid := []Line{
		{"aaaabbbbcccc", 1}, {"ddddeeeeffff", 2}, {"gggghhhhiiii", 3},
	}
	win := windowLines(grid, View{Row: 1, Col: 4}, 2, 4)
	if len(win) != 2 || win[0].Text != "eeee" || win[0].Color != 2 || win[1].Text != "hhhh" || win[1].Color != 3 {
		t.Fatalf("got %+v", win)
	}
}

func TestAnchorRow(t *testing.T) {
	if anchorRow(10, 8) != 4 {
		t.Fatal("anchor")
	}
	if anchorRow(1, 8) != 0 || anchorRow(0, 8) != 0 {
		t.Fatal("floor")
	}
}

func TestWindowLinesEnd(t *testing.T) {
	grid := []Line{{"x", 9}}
	if len(windowLines(grid, View{Row: 0, Col: 0}, 5, 10)) != 1 {
		t.Fatal("end")
	}
}

File: server/test/viewport.test.js
'use strict';
const { test } = require('node:test');
const assert = require('node:assert');
const { panViewport, windowLines, anchorRow, ROW_STEP, COL_STEP } = require('../lib/viewport');

// The server owns an omnidirectional viewport over the terminal's NATIVE-width
// grid; the device forwards arrow keys and renders the window the server sends.

test('panViewport up/down moves rows and unsticks follow', () => {
  assert.deepEqual(panViewport({ row: 5, col: 0, follow: true }, 'up'),
    { row: 5 - ROW_STEP, col: 0, follow: false });
  assert.deepEqual(panViewport({ row: 5, col: 0, follow: false }, 'down'),
    { row: 5 + ROW_STEP, col: 0, follow: false });
});

test('panViewport left/right moves columns, floored at 0', () => {
  assert.equal(panViewport({ row: 0, col: 10, follow: false }, 'left').col, 10 - COL_STEP);
  assert.equal(panViewport({ row: 0, col: 0, follow: false }, 'left').col, 0);   // floored
  assert.equal(panViewport({ row: 0, col: 0, follow: false }, 'right').col, COL_STEP);
});

test('panViewport floors row at 0 and ignores unknown keys', () => {
  assert.equal(panViewport({ row: 0, col: 0, follow: false }, 'up').row, 0);
  assert.deepEqual(panViewport({ row: 2, col: 2, follow: true }, 'zzz'), { row: 2, col: 2, follow: true });
});

test('windowLines slices rows x cols at the viewport offset, keeping colors', () => {
  const grid = [
    { text: 'aaaabbbbcccc', color: 1 },
    { text: 'ddddeeeeffff', color: 2 },
    { text: 'gggghhhhiiii', color: 3 },
  ];
  const win = windowLines(grid, { row: 1, col: 4 }, { rows: 2, cols: 4 });
  assert.deepEqual(win, [
    { text: 'eeee', color: 2 },
    { text: 'hhhh', color: 3 },
  ]);
});

test('anchorRow places the pointer near the bottom so the question above is visible', () => {
  assert.equal(anchorRow(10, 8), 10 - (8 - 2));  // pointer on view row 6 of 8
});

test('anchorRow floors at 0 when the pointer is near the top of the grid', () => {
  assert.equal(anchorRow(1, 8), 0);
  assert.equal(anchorRow(0, 8), 0);
});

test('windowLines stops at the end of the grid (no padding rows)', () => {
  const grid = [{ text: 'x', color: 9 }];
  const win = windowLines(grid, { row: 0, col: 0 }, { rows: 5, cols: 10 });
  assert.equal(win.length, 1);
});
